src/main/scala/mesosphere/marathon/tasks/PortsMatcher.scala

```diff
@@ -47,145 +53,224 @@ class PortsMatcher(app: AppDefinition, offer: Offer, acceptedResourceRoles: Set[
+      protos.Range(port.toLong, port.toLong)
+    }
+  }
+
+  object PortWithRole {
+    /**
+      * Return RangesResources covering all given ports with the given roles.
+      *
+      * Creates as few RangesResources as possible while
+      * preserving the order of the ports.
+      */
+    def createPortsResources(resources: Seq[PortWithRole]): Seq[RangesResource] = {
+      /*
+       * Create as few ranges as possible from the given ports while preserving the order of the ports.
+       *
+       * It does not check if the given ports have different roles.
+       */
+      def createRanges(ranges: Seq[PortWithRole]): Seq[protos.Range] = {
+        val builder = Seq.newBuilder[protos.Range]
+
+        @tailrec
+        def process(lastRangeOpt: Option[protos.Range], next: Seq[PortWithRole]): Unit = {
+          (lastRangeOpt, next.headOption) match {
+            case (None, _) =>
+            case (Some(lastRange), None) =>
+              builder += lastRange
+            case (Some(lastRange), Some(nextPort)) if lastRange.end == nextPort.port - 1 =>
+              process(Some(lastRange.copy(end = nextPort.port.toLong)), next.tail)
+            case (Some(lastRange), Some(nextPort)) =>
+              builder += lastRange
+              process(Some(nextPort.toRange), next.tail)
           }
+        }
+        process(ranges.headOption.map(_.toRange), ranges.tail)
+
+        builder.result()
+      }
+
+      val builder = Seq.newBuilder[RangesResource]
+      @tailrec
+      def process(resources: Seq[PortWithRole]): Unit = resources.headOption match {
+        case None =>
+        case Some(PortWithRole(role, _)) =>
+          val portsForResource: Seq[PortWithRole] = resources.takeWhile(_.role == role)
+          builder += RangesResource(name = Resource.PORTS, createRanges(portsForResource), role = role)
+          process(resources.drop(portsForResource.size))
+      }
+      process(resources)
+
+      builder.result()
+    }
+
+    /**
+      * We want to make it less likely that we are reusing the same dynamic port for tasks of different apps.
+      * This way we allow load balancers to reconfigure before reusing the same ports.
+      *
+      * Therefore we want to choose dynamic ports randomly from all the offered port ranges.
+      * We want to use consecutive ports to avoid excessive range fragmentation.
+      *
+      * The implementation idea:
+      *
+      * * Randomize the order of the offered ranges.
+      * * Now treat the ports contained in the ranges as one long sequence of ports.
+      * * We randomly choose an index where we want to start assigning dynamic ports in that sequence. When
+      *   we hit the last offered port with wrap around and start offering the ports at the beginning
+      *   of the sequence up to (excluding) the port index we started at.
+      */
+    def lazyRandomPortsFromRanges(rand: Random = Random)(offeredPortRanges: Seq[PortRange]): Iterator[PortWithRole] = {
```

Comment:
First parameter list is conceptually for an initialization state. You could wire the rand via dependency injection and just directly use the function with the second parameter list. The second parameter list is for parameters that are specific to the invocation.